plugins/omdtls/omdtls.c

```diff
@@ -0,0 +1,885 @@
+		char *line;
+		line = X509_NAME_oneline(X509_get_subject_name(cert), 0, 0);
+		dbgprintf("dtls_connect[%p]: Subject: %s\n", pWrkrData, line);
+		OPENSSL_free(line);
+	
+		line = X509_NAME_oneline(X509_get_issuer_name(cert), 0, 0);
+		dbgprintf("dtls_connect[%p]: Issuer: %s\n", pWrkrData, line);
+		OPENSSL_free(line);
+		X509_free(cert);
+	} else {
+		dbgprintf("dtls_connect[%p]: No certificates.\n", pWrkrData);
+	}
+*/
+
+	/* Set reference in SSL obj */
+	SSL_set_ex_data(pWrkrData->sslClient, 0, NULL);
+	SSL_set_ex_data(pWrkrData->sslClient, 1, NULL);
+
+	/* Set and activate timeouts */
+	struct timeval timeout;
+	timeout.tv_sec = 3;
+	timeout.tv_usec = 0;
+	BIO_ctrl(bio_client, BIO_CTRL_DGRAM_SET_RECV_TIMEOUT, 0, &timeout);
+
+
+	pWrkrData->bIsConnected = 1;
+finalize_it:
+	DBGPRINTF("dtls_connect[%p]: Exit with %d for %s:%s\n", pWrkrData, iRet, pData->target, pData->port);
+	if(iRet != RS_RET_OK) {
+		pWrkrData->bIsConnected = 0;
+	}
+	RETiRet;
+}
+
+/*
+*	Setup socket and establish virtual Connect on a SOCKET Level for DTLS 
+*/
+static rsRetVal
+dtls_init(wrkrInstanceData_t *pWrkrData)
+{
+	DEFiRet;
+	int iErr;
+	instanceData *const pData = (instanceData *const) pWrkrData->pData;
+	struct addrinfo *addrResolved = NULL;
+	struct addrinfo hints;
+	DBGPRINTF("dtls_init[%p]: setup for %s:%s\n", pWrkrData, pData->target, pData->port);
+
+	// Close Before
+	dtls_close(pWrkrData);
+
+	// Setup receiving Socket for DTLS
+	if((pWrkrData->sockin = socket(AF_INET, SOCK_DGRAM, 0)) == -1)
+		return 1;
+	memset(&pWrkrData->dtls_client_addr, 0, sizeof(pWrkrData->dtls_client_addr));
+	pWrkrData->dtls_client_addr.sin_family = AF_INET;
+	pWrkrData->dtls_client_addr.sin_port = htons(0);
+	pWrkrData->dtls_client_addr.sin_addr.s_addr = INADDR_ANY;
+	if (bind(pWrkrData->sockin,
+		(struct sockaddr*)&pWrkrData->dtls_client_addr,
+		sizeof(pWrkrData->dtls_client_addr)) < 0) {
+		LogError(0, RS_RET_SUSPENDED, "omdtls[%p]: dtls_init Unable to bind DTLS CLient socket with errno %d",
+			pWrkrData, errno);
+		ABORT_FINALIZE(RS_RET_COULD_NOT_BIND);
+	}
+
+	// Resolve target address
+	memset(&hints, 0, sizeof(hints));
+	/* port must be numeric, because config file syntax requires this */
+	hints.ai_family = glbl.GetDefPFFamily(runModConf->pConf);
+	hints.ai_socktype = SOCK_DGRAM;
+	if((iErr = (getaddrinfo((char*)pData->target, (char*)pData->port, &hints, &addrResolved))) != 0) {
+		LogError(0, RS_RET_SUSPENDED,
+			"omdtls[%p]: could not get addrinfo for hostname '%s':'%s': %s",
+			pWrkrData, pData->target, pData->port, gai_strerror(iErr));
+		ABORT_FINALIZE(RS_RET_ADDRESS_UNKNOWN);
+	}
+	pWrkrData->dtls_rcvr_addrinfo = addrResolved;
+	addrResolved = NULL;
+
+	// Init Socket Connection (Which technically does not connect but prepares socket for DTLS)
+	DBGPRINTF("dtls_init[%p]: Init Session to %s:%s\n", pWrkrData, pData->target, pData->port);
+
+	pWrkrData->sockout = socket(AF_INET, SOCK_DGRAM, 0);
+	// Connect the UDP socket to the server's address
```

Comment:
reconsider comment? We are acting as client here, not as server. Is the comment or the code wrong?

Reply:
Actually this is not true. As we are in connectionless UDP mode, we need an incoming and outgoing UDP Socket for sending and receiving. The handling is done inside the OpenSSL library.

Reply:
The word "server" trigerred me. It has the co-notation of being used only in the server case. I strongly suggest to connect that comment with what you wrote here in the conversation. That's a much better and useful description. And it's not obvious if you only read the code.

Reply:
Alright I will change the comment